### DISCUSSION TITLE
implement Comments

### DISCUSSION THREAD

**Discussion by Remian103** [Remian103/lolskin-price-tracker#47] [Ideas]
### 구현 목표
1. 댓글 기능
2. 댓글마다의 좋아요 기능
-----------------------------------
### 고려사항

1. 댓글 리스트를 불러오는 방식
2. 좋아요 횟수를 업데이트 하는 방식 (like, unlike)

#### 댓글 리스트 불러오는 방식
1. 목록 불러오는 API
    - protocol (ex. http get)
    - 전체 댓글 리스트 불러오기
    - 페이지 단위로 불러오기
    - 스크롤이 내려갈 때마다 추가로 불러오기
2. 목록 업데이트 방식
    - 별도의 refresh 버튼을 통해 업데이트
    - 유저가 댓글을 submit 했을 때 목록 업데이트
    - 실시간 업데이트 (다른 브라우저에서 댓글이 추가될 때도 업데이트) - 아마 web socket이 필요할 것 같음
3. 댓글 submit API
    - http post
      - 요청 형식, 응답 형식

#### 좋아요 횟수에 대한 방식
1. API (ex. http post method)
    - 요청 형식, 응답 형식
2. 좋아요 횟수 업데이트 방식
    - 좋아요 버튼을 눌렀을 때 해당 댓글의 좋아요 수만 업데이트
    - 좋아요 버튼을 눌렀을 때 모든 댓글의 좋아요 업데이트  - 댓글 목록 자체를 refresh하는 형태로 구현될 듯
    - 실시간 업데이트

**Comment by Remian103** [comment]
_2021.08.26 삭제됨_
_해당 내용 최상단으로 이동됨_

**Reply by Remian103** [reply]
일단 목록에서 상위에 있는 것들이 당장 구현하기 쉬울 듯....?

실시간 업데이트는 대강 아래 사진같은 느낌? 근데 web socket 따로 또 공부해야 가능할듯
![실시간 좋아요](https://images.contentful.com/1es3ne0caaid/uFNPM8Sl7Eg6Y2U0OUYyA/5a280590f9c939393a1e3cf3b0ae0494/realtime-likes-nodejs-demo.gif)

**Comment by jiyolla** [comment]
아~ 좋아요

1번은 endpoint하나로 처리할 수 있을듯?
/api/skins/{skin_id}/comments/?skip=0&limit=100
skip이 가져올 첫번째 댓글, limit이 마지막 댓글
skip limit,기본값 설정 가능.

페이지 단위, 스크롤 단위는 컨트롤은 프론트엔드에서 하는 게 좋을 수도? 한 페이지/스크롤에 몇 개가 들어갈지는 그 스크린 사이즈 댓글 길이 등에 따라 가변하는 게 시각적으로 좋을 수도 있어서

순서가 이제 좀 살짝 문제가 될 수 있는데(특히 websocket으로 할 때 실시간으로 순서가 변화할 때), 기본적으로는 좋아요 내림차순.

**Reply by jiyolla** [reply]
댓글 가능의 1번에 대한 리뷰

**Reply by Remian103** [reply]
api에서 skip, limit을 댓글 array를 자르는 인덱스 정도로 생각하면 되는건가..??
댓글 정보 넘어올 때는 전체 댓글의 수도 같이 넘어오는 게 좋을듯

그리고 댓글 정렬 순서는
만약 댓글이 전체가 아니라 부분적으로 불러와지는 형태면 시간순으로 나열하는게 자연스러울 거 같은데
좋아요를 많이 받은 top3 best 댓글 같은 형식도 생각해볼만 한듯?

**Reply by jiyolla** [reply]
ㅇㅎ 시간 순으로 가져오면, 보여줄 때도 시간 순으로 보여주는 건가?

**Reply by jiyolla** [reply]
아니면 가져와서 FE에서 자체 정렬하는 건가?

**Reply by Remian103** [reply]
skip limit을 정해서 요청하는 형태면 be에서 정렬해야 되는거 아님? skip limit의 기준이 있을텐데

**Comment by jiyolla** [comment]
나머지 부분은 다 괜찮은듯

API는

### 한 스킨의 댓글 가져오기
get /api/skins/{skin_id}/comments/?skip={first_comment_idx}&limit={last_comment_idx}
여기다가 댓글 메타 정보(총 개수 등)를 추가와 요청한 구간의 댓글 들을 리스트로 반환하면 될 듯.

지금 당장 구체화 좀 해보자면
```
Comment_meta  ## Not a good name
[
  'skin_id':  int
  'num_comments': int
  'first_comment_idx': int
  'last_comment_idx': int   ## It could be different from the one from the query, as there might not be that much comments
  'comments': List[Comment]
  ...
]

Comment
[
  'comment_id': int
  'created': date
  'modified': date
  'author_id': int  ## If authentication is available
  'content': str
  'likes': int
  'dislikes': int
  ...
]

```
### 좋아요/싫어요 서버에 송신하기
post /api/comments/{comment_id}/likes
post /api/comments/{comment_id}/dislikes
인증이 생기 전까지는, request body비어도 괜찮을 듯

인증이 생긴 후에는
delete /api/comments/{comment_id}/likes
delete /api/comments/{comment_id}/dislikes
도 추가하여, 좋아요/싫어요 취소 기능도 추가.
